File: languages/cpp/03-advanced/structs-and-classes/example/class-with-validation.cpp
#include <iostream>
#include <stdexcept>
#include <string>

using namespace std;

class BankAccount {
public:
    BankAccount(string ownerName, double initialBalance)
        : owner(ownerName), balance(0.0) {
        if (owner.empty()) {
            throw invalid_argument("Owner name cannot be empty.");
        }
        if (initialBalance < 0.0) {
            throw invalid_argument("Initial balance cannot be negative.");
        }
        balance = initialBalance;
    }

    bool deposit(double amount) {
        if (amount <= 0.0) {
            return false;
        }
        balance += amount;
        return true;
    }

    bool withdraw(double amount) {
        if (amount <= 0.0 || amount > balance) {
            return false;
        }
        balance -= amount;
        return true;
    }

    const string& getOwner() const {
        return owner;
    }

    double getBalance() const {
        return balance;
    }

private:
    string owner;
    double balance;
};

int main() {
    try {
        BankAccount account("Ana Smith", 100.0);

        cout << "Account owner: " << account.getOwner() << '\n';
        cout << "Initial balance: " << account.getBalance() << '\n';

        if (account.deposit(50.0)) {
            cout << "Deposit successful.\n";
        }

        if (account.withdraw(30.0)) {
            cout << "Withdraw successful.\n";
        }

        if (!account.withdraw(500.0)) {
            cout << "Withdraw rejected (insufficient funds).\n";
        }

        cout << "Final balance: " << account.getBalance() << '\n';
    } catch (const exception& ex) {
        cout << "Error: " << ex.what() << '\n';
    }

    return 0;
}

File: languages/go/03-advanced/structs-and-classes/example/class-with-validation.go
// This helper example focuses on keeping validation close to the type that owns the data.

// This extra example extends structs-and-classes with validated object state.

package main

import (
	"errors"
	"fmt"
)

// Keep this helper separate so the main example can focus on the larger idea without extra noise.
type bankAccount struct {
	owner   string
	balance float64
}

func newBankAccount(owner string, initialBalance float64) (*bankAccount, error) {
	if owner == "" {
		return nil, errors.New("owner name cannot be empty")
	}
	if initialBalance < 0.0 {
		return nil, errors.New("initial balance cannot be negative")
	}

	return &bankAccount{
		owner:   owner,
		balance: initialBalance,
	}, nil
}

func (account *bankAccount) deposit(amount float64) bool {
	if amount <= 0.0 {
		return false
	}
	account.balance += amount
	return true
}

func (account *bankAccount) withdraw(amount float64) bool {
	if amount <= 0.0 || amount > account.balance {
		return false
	}
	account.balance -= amount
	return true
}

func main() {
	account, err := newBankAccount("Ana Smith", 100.0)
	if err != nil {
		fmt.Printf("Error: %s\n", err)
		return
	}

	fmt.Printf("Account owner: %s\n", account.owner)
	fmt.Printf("Initial balance: %v\n", account.balance)

	if account.deposit(50.0) {
		fmt.Println("Deposit successful.")
	}
	if account.withdraw(30.0) {
		fmt.Println("Withdraw successful.")
	}
	if !account.withdraw(500.0) {
		fmt.Println("Withdraw rejected (insufficient funds).")
	}

	fmt.Printf("Final balance: %v\n", account.balance)
}
